### PR TITLE
feat(moe): wire in-kernel routing (FromLogits) into unified MoE API + routing axes in fuzzer

### PR DESCRIPTION
**Pull request by feih-nv**
## 📌 Description
Continues #3590 by @YangXu1990uiuc, preserving the original authorship.
### Why
The unified `MoELayer` API currently accepts only precomputed routing. Its fuzzer therefore cannot exercise the TRTLLM in-kernel routing path, including regressions involving:
- #2796 — FP32 router logits
- #2822 — all-negative logits
- #2575 — DeepSeek-V3 grouped routing
- #2485 / #2907 — routing bias and weight handling

This PR exposes the existing TRTLLM `FromLogits` implementation through the unified API; it adds no new CUDA kernels.
#### Routing API and dispatch
`MoEActivationPack` now supports two mutually exclusive routing modes:
- **Pre-routed (default):** `topk_ids` and `topk_weights` are required;
  `routing_logits` and `routing_bias` must be absent.
- **From logits:** `routing_logits` is required and `routing_bias` is optional;
  `topk_ids` and `topk_weights` must be absent. The TRTLLM kernel computes
  top-k selection in-kernel.

These contracts are validated at pack construction and again before launch.

The new mode/logits/bias fields are keyword-only, preserving the existing
positional constructor. Pre-routed fields are renamed
`selected_experts` → `topk_ids` and `final_scales` → `topk_weights`;
positional callers remain compatible, while keyword callers must update.

Runners declare supported routing modes. `MoELayer` filters candidates and
caches winners by `(token_bucket, routing_mode)`, preventing logits input from
reaching pre-routed-only backends such as CuteDSL.

`RoutingInputMode` moves to `flashinfer/tllm_enums.py` with the other kernel ABI
enums and remains re-exported from `fused_moe.core`.
#### Validation and robustness
- Validates routing mode/field consistency when constructing an activation pack.
- Revalidates routing tensors at the runner boundary:
  - `topk_ids/topk_weights`: `[num_tokens, top_k]`
  - `routing_logits`: `[num_tokens, num_experts]`, FP32 or BF16
  - `routing_bias`: `[num_experts]`, FP32 or BF16
  - precomputed expert IDs: `torch.int32`
  - all pack tensors on the activation device
- Allocates kernel-produced expert weights as BF16, independently of logits dtype.
- Fixes the unified NVFP4 benchmark’s global expert IDs and clamped top-k geometry.
- Applies the fuzzer memory budget after final routing/EP decisions.
- Makes printed reproduction commands shell-safe.
- Documents that `MoELayer` instances are not thread-safe because runners retain per-call launch state.
#### Routing-aware fuzzer
Adds the following axes to `test_unified_moe_fuzz.py`:
- routing input mode: pre-routed / `FromLogits`
- all `RoutingMethodType` methods
- FP32 and BF16 logits
- routing bias
- DeepSeek-V3 group parameters
- all-negative and skewed routing distributions

The same host routing oracle drives the expected result for both routing modes. Curated seeds cover the headline regressions, and `FLASHINFER_UMOE_FUZZ_ONLY_SEED` provides deterministic reproduction.

### ⚠️ Compatibility
The field rename is intentionally breaking for keyword callers; no deprecated
aliases are included because the unified API is still pre-stable.

### Out of scope
- BF16 runner `FromLogits`
- `FromLogits` with expert parallelism
- `UnpackedPrecomputed` unified runners
- routing replay through the unified API
- stateless/thread-safe runner calls
- enabling the unified fuzzer in CI

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## 🧪 Tests

**Validation (SM100 Blackwell):**

| Suite | Result |
|---|---|
| `tests/moe/test_unified_moe.py` + `tests/moe_ep/test_compute_bridge.py` (incl. #3867 BF16 conformance) | 154 passed, 1 skipped |
| Curated FromLogits fuzzer seeds (DSv3 grouped+bias, fp32 logits, all-negative) | 4 passed |
| CPU config-contract tests | 147 passed, 1 skipped |

The unified fuzzer remains opt-in and is not enabled in normal CI.

## Reviewer Notes
Two decisions need explicit maintainer sign-off:
1. The intentional breaking field rename without compatibility aliases.
2. Reusing the host routing oracle ported from the existing TRTLLM routing references.

## Summary by CodeRabbit

* **New Features**
  * Added configurable routing modes for Mixture-of-Experts execution, including routing directly from logits.
  * Exposed `RoutingInputMode` through the public API.
  * Added support for routing biases and broader routing configurations.

* **Bug Fixes**
  * Improved validation for routing data types, shapes, modes, and device placement.
  * Prevented incompatible execution strategies from being selected or reused.
  * Updated NVFP4 and expert-parallel routing to preserve correct expert identifiers and weights.

* **Tests**
  * Expanded coverage for logits-based routing, validation, fuzzing, determinism, and output-buffer correctness.

### CODE REVIEW

File: flashinfer/fused_moe/runners.py [hunk 1]
```diff
@@ -255,30 +285,71 @@ def pack_inputs(
...
+                "FromLogits computes topk_ids/topk_weights in-kernel; leave them None."
+            )
+            # Validate here: the launcher enforces this with a C++ ICHECK whose
+            # failure message doesn't name the offending Python-side input.
+            assert act.routing_logits.dtype in (torch.float32, torch.bfloat16), (
+                f"routing_logits must be float32 or bfloat16, "
+                f"got {act.routing_logits.dtype}."
+            )
+            routing_logits = act.routing_logits
+            routing_bias = act.routing_bias
```
[the remark this answers is not in the document]
Comment (feih-nv) [reply flashinfer/fused_moe/runners.py:407]: Partially adopted in 01ce3ec. The dtype-equality check is intentionally not taken: the launcher derives `mRoutingBiasDtype` and `mRoutingLogitsDtype` independently and mixed combinations are supported (`test_routing_dtype_flexibility` parametrizes them separately; fp32 logits + bf16 bias is the standard DeepSeek-V3 shape). What was missing is a dtype membership check -- the launcher maps ->Bfloat16 and anything else->Fp32 with no ICHECK, so an fp16 bias would be silently misread; `pack_inputs` now rejects non-{bf16, fp32} bias. The shape check is already enforced by the C++ launcher (`check_routing_bias_shape`).

File: flashinfer/fused_moe/api.py [hunk 2]
```diff
@@ -499,12 +499,44 @@ def __getitem__(self, key: str):
...
+    # call site fails loudly instead of silently binding a tensor to the mode.
+    routing_input_mode: RoutingInputMode = field(
+        default=RoutingInputMode.PackedPrecomputed, kw_only=True
+    )
+    routing_logits: Optional[Tensor] = field(
+        default=None, kw_only=True
+    )  # [M, num_experts] float32 or bfloat16
+    routing_bias: Optional[Tensor] = field(
+        default=None, kw_only=True
+    )  # [num_experts] bfloat16 or float32 (independent of logits dtype)
```
[the remark this answers is not in the document]
Comment (feih-nv) [reply flashinfer/fused_moe/api.py:570]: FlashInfer requires Python 3.10+.

File: flashinfer/fused_moe/runners.py [hunk 3]
```diff
@@ -95,6 +97,18 @@ def pack_inputs(
...
+        # logits pack's None topk tensors into the kernel launch.
+        if act.routing_input_mode not in self.supported_routing_modes:
+            raise NotImplementedError(
+                f"CuteDslNvfp4Runner does not support "
+                f"routing_input_mode={act.routing_input_mode!r} "
+                "(only PackedPrecomputed is wired; CuteDSL has no in-kernel router)."
+            )
+        assert act.topk_ids is not None and act.topk_weights is not None, (
+            "routing_input_mode=PackedPrecomputed requires topk_ids + topk_weights."
+        )
```
[the remark this answers is not in the document]
Comment (feih-nv) [reply flashinfer/fused_moe/runners.py]: Adopted in 55bbae04c, centralized rather than per-site: MoEActivationPack.__post_init__ validates that all routing tensors (topk_ids, topk_weights, routing_logits, routing_bias) are on hidden_states_q's device, covering every runner path from one place. (Resolving the sibling comments on the other lines as duplicates of this.)

File: flashinfer/fused_moe/runners.py [hunk 4]
```diff
@@ -257,30 +287,87 @@ def pack_inputs(
...
+            # mRoutingBiasDtype and mRoutingLogitsDtype separately; mixed
+            # combinations are supported — see test_routing_dtype_flexibility).
+            # But the launcher maps bf16 -> Bfloat16 and ANYTHING ELSE -> Fp32
+            # with no dtype ICHECK, so e.g. an fp16 bias would be silently
+            # reinterpreted as fp32 bits. Reject non-{bf16, fp32} here.
+            if act.routing_bias is not None:
+                assert act.routing_bias.dtype in (torch.bfloat16, torch.float32), (
+                    f"routing_bias must be bfloat16 or float32, "
+                    f"got {act.routing_bias.dtype}."
+                )
```
[the remark this answers is not in the document]
Comment (feih-nv) [reply flashinfer/fused_moe/runners.py]: ditto

File: flashinfer/fused_moe/runners.py [hunk 5]
```diff
@@ -293,8 +380,8 @@ def pack_inputs(
         # MoERunner.forward.  None-valued entries are the optional gemm bias /
         # swiglu beta-clamp / per-token-scale paths not used by the MVP.
         self._static_kwargs = dict(
-            routing_input_mode=RoutingInputMode.PackedPrecomputed,
-            routing_bias=None,
+            routing_input_mode=routing_input_mode,
+            routing_bias=routing_bias,
```
[the remark this answers is not in the document]
Comment (feih-nv) [reply flashinfer/fused_moe/runners.py:464]: Valid observation, but the pattern predates this PR — _static_kwargs shipped with the original TrtllmFp4RoutedRunner (#3093); this PR adds two keys to the existing dict. The autotuner's pack_inputs → choose_one → forward contract is inherently sequential, so single-threaded use (including autotune) is safe. Adopted the documentation option in 55bbae04c: MoELayer's docstring now states the not-thread-safe contract (one layer per thread/stream) and names the stateless calling convention as a tracked follow-up. Refactoring the calling convention is out of scope here.

File: flashinfer/fused_moe/api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (feih-nv) [reply flashinfer/fused_moe/api.py:570]: Adopted in 55bbae04c — __post_init__ validates mode↔field consistency (logits ⟺ FromLogits, topk pair ⟺ pre-routed, bias only with logits), topk_ids dtype, and device consistency at construction time, with raises rather than asserts.

File: flashinfer/fused_moe/runners.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (feih-nv) [reply flashinfer/fused_moe/runners.py]: Partially adopted in 55bbae04c: all new user-input validation now lives in MoEActivationPack.__post_init__ using ValueError/TypeError raises, which survive python -O. The module-internal asserts in pack_inputs are kept as-is for consistency with the established convention in this module (core.MoERunner.forward and the canonical wrappers use the same style) — a repo-wide policy change belongs in its own PR.

**Review by aleozlx — APPROVED** [review]

**Comment by feih-nv** [comment]
@coderabbitai review

File: tests/moe/test_unified_moe.py [hunk 6]
```diff
@@ -609,6 +609,214 @@ def test_non_swiglu_activation_rejected(self, act):
...
+            MoEActivationPack(x, **fields)
+
+
+class TestRunnerBoundaryValidation:
+    """The shared ``_validate_*`` helpers, called directly (CPU, no JIT).
+
+    They duplicate ``__post_init__`` BY DESIGN: the pack is mutable, so the
+    launch boundary is the authoritative validation layer. The mutation tests
+    below pin exactly the bypass that motivates the duplication -- do not
+    "deduplicate" these checks against ``__post_init__``.
```
[the remark this answers is not in the document]
Comment (feih-nv) [reply tests/moe/test_unified_moe.py:723]: We already cover device mismatches in TestActivationPackValidation (meta-tensor device mismatch tests).

File: flashinfer/fused_moe/runners.py [hunk 7]
```diff
@@ -257,30 +390,64 @@ def pack_inputs(
...
+            # the offset, which the kernel treats as non-local and skips → zero output.
+            _validate_prerouted_inputs(
+                act, num_tokens, routing.top_k, "TrtllmFp4RoutedRunner"
+            )
+            routing_logits = None
+            routing_bias = None
+            ids = act.topk_ids
+            weight_bf16_bits = (
+                act.topk_weights.to(torch.bfloat16).view(torch.int16).to(torch.int32)
+            )
```
[the remark this answers is not in the document]
Comment (feih-nv) [reply flashinfer/fused_moe/runners.py:433]: Thanks for the review — .view() for same-element-size dtype reinterpretation (bf16 ↔ int16, both 2 bytes) does not require contiguity in PyTorch; it reinterprets each element's bytes in-place with strides unchanged. I've verified across 4 non-contiguous tensor shapes including irregular strides (PyTorch 2.9.0).

File: flashinfer/fused_moe/runners.py [hunk 8]
```diff
@@ -257,30 +390,64 @@ def pack_inputs(
         if hidden_states_scale.dtype == torch.uint8:
             hidden_states_scale = hidden_states_scale.view(torch.float8_e4m3fn)
```
[the remark this answers is not in the document]
Comment (feih-nv) [reply flashinfer/fused_moe/runners.py:391]: Ditto to above

File: flashinfer/fused_moe/runners.py [hunk 9]
```diff
@@ -447,20 +616,29 @@ def pack_inputs(
...
         # the offset, which the kernel treats as non-local and skips → zero output.
-        ids = act.selected_experts
+        _validate_prerouted_inputs(
+            act, num_tokens, routing.top_k, "TrtllmBf16RoutedRunner"
+        )
+        ids = act.topk_ids
         weight_bf16_bits = (
-            act.final_scales.to(torch.bfloat16).view(torch.int16).to(torch.int32)
+            act.topk_weights.to(torch.bfloat16).view(torch.int16).to(torch.int32)
         )
```
[the remark this answers is not in the document]
Comment (feih-nv) [reply flashinfer/fused_moe/runners.py:637]: Ditto.
